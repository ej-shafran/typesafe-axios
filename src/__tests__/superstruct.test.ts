import { taxios } from "..";

import {
  array,
  literal,
  object,
  optional,
  string,
  union,
  Infer,
} from "superstruct";
import nock from "nock";
import * as UTILS from "./__utils__";

const SCHEMA = object({
  id: string(),
  username: string(),
  email: string(),
  address: object({
    street: string(),
    zipcode: optional(string()),
  }),
  followers: array(
    object({
      id: string(),
      role: optional(union([literal("admin"), literal("guest")])),
    })
  ),
});

type TYPE = Infer<typeof SCHEMA>;
function ensure(_: TYPE) { }

describe("superstruct interop", () => {
  it("returns `data` and not `error` when the data is valid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.VALID);

    const response = await taxios(SCHEMA).get(UTILS.API);
    const [error, data] = response.data;

    expect(error).toBeUndefined();
    expect(data).toEqual(UTILS.VALID);
    ensure(data!);
  });

  it("returns `error` and not `data` when the data is invalid", async () => {
    nock(UTILS.API).get("/").reply(200, UTILS.INVALID);

    const response = await taxios(SCHEMA).get(UTILS.API);
    const [error, data] = response.data;

    expect(data).toBeUndefined();
    expect(JSON.stringify(error, null, 2)).toMatchSnapshot();
  });
});
