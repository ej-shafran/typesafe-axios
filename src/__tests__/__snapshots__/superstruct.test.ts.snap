// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`superstruct interop returns \`error\` and not \`data\` when the data is invalid 1`] = `
"{
  "type": "string",
  "key": "email",
  "path": [
    "email"
  ],
  "branch": [
    {
      "id": "SOME_ID",
      "username": "SOME_USERNAME",
      "address": "ABC123 St.",
      "followers": []
    },
    null
  ],
  "name": "StructError"
}"
`;
